### PR TITLE
Estimator size

### PR DESCRIPTION
**Pull request by charlessolar**
Added a property Size to estimator to report how many hashes have been added to the object

### CODE REVIEW

File: CardinalityEstimation.Test/CardinalityEstimatorTests.cs [hunk 1]
```diff
@@ -65,6 +65,25 @@ public void TestGetSigma()
         }
 
         [TestMethod]
+        public void TestSize()
+        {
+            var estimator = new CardinalityEstimator();
+
+            Assert.AreEqual((UInt64)0, estimator.Size);
```
Comment (oronnavon) [inline CardinalityEstimation.Test/CardinalityEstimatorTests.cs]: use 0UL (and same below)

File: CardinalityEstimation.Test/CardinalityEstimatorTests.cs [hunk 2]
```diff
@@ -65,6 +65,25 @@ public void TestGetSigma()
         }
```
Comment (oronnavon) [inline CardinalityEstimation.Test/CardinalityEstimatorTests.cs:65]: Welcome, and thanks for your contribution!

File: CardinalityEstimation/ICardinalityEstimator.cs [hunk 3]
```diff
@@ -42,5 +42,7 @@ public interface ICardinalityEstimator<in T>
         /// </summary>
         /// <returns>The estimated count of unique elements</returns>
         ulong Count();
+
+        ulong Size { get; }
```
Comment (oronnavon) [inline CardinalityEstimation/ICardinalityEstimator.cs]: Let's rename this, how about "CountElementsAdded" ? Also, add the following comment:

```
/// <summary>
///     Returns the number of times elements were added (including duplicates)
/// </summary>
/// <returns>The number of times <see cref="Add"/> was called</returns>
ulong CountElementsAdded { get; }
```

File: CardinalityEstimation.Test/CardinalityEstimatorTests.cs (hunk 2 above)
Comment (oronnavon) [inline CardinalityEstimation.Test/CardinalityEstimatorTests.cs:65]: Note that the new property is not serialized by the custom serializer. However, that's fine - I'm planning to merge the Murmur3 branch soon, and it includes a breaking change in the serializer.  So go ahead and ignore the serializer for now.

**Comment by charlessolar** [comment]
Ok - renamed the property and added the serialization.  
Had to move CountElementsAdded into each .Add method because the initialization from state uses AddElementHash which I guess is why I didn't notice the size not being serialized before.. oops

File: CardinalityEstimation.Test/CardinalityEstimatorTests.cs (hunk 2 above)
Comment (oronnavon) [inline CardinalityEstimation.Test/CardinalityEstimatorTests.cs:65]: The point of serializedDense_v1.0.bin, serializedDirect_v1.0.bin and serializedSparse_v1.0.bin is to make sure we don't introduce breaking changes without incrementing the major version - these files should not be changed :)  Rather, we need to support backwards compatibility.

As I mentioned, I already have a serializer change in the works (in the Murmur3 branch), so you can just drop the serializer change here, and I'll make sure the CountElementsAdded is serialized as well in my next merge.

Thanks,
Oron

**Comment by charlessolar** [comment]
Ok no problem, I removed the serializer changes and tests.  
It raises an interesting question though - I am using the serializer to store the estimators in a db while calculating ongoing cardinality in my data.  Are you planning on supporting the old serialized formats alongside the new one?  In the code base currently if the version doesn't match its not dealt with.

Once this new stuff comes out if need be I'll write some backwards compatibility in a PR because for my use case having a database full of suddenly unreadable estimators would be a disaster haha

**Comment by oronnavon** [comment]
Don't worry, the new Murmur3 branch has backward compatibility. You can have a look here:
https://github.com/Microsoft/CardinalityEstimation/tree/murmur3

**Pull request merged**